# Mini-Project I
During this project, we will practice handling of complex lists and dictionaries in Python. Plus, we will learn how to work with API documentation. Don't be afraid to search for the information in the [**documentation**](https://api.tfl.gov.uk/swagger/ui/index.html?url=/swagger/docs/v1#!/AccidentStats/AccidentStats_Get).

Let's go to the tasks, we have some parsing to do :)!!

In [1]:
# import packages we need (remember what packages we used yesterday during the API session)
# IMPORT HERE

import requests as re
import os # ?
import json
import config as cfg


## Demo

In [2]:

app_id = cfg.tfl['PK']
app_key = cfg.tfl['SK']


# # don't forget package os
# app_id = os.environ["<your app id name>"]
# app_key = os.environ["<your app key name>"]
url_append = f'?app_id={app_id}&app_key={app_key}' 

In [3]:
print(url_append)

?app_id=b09abd2194b44c4a8f80e810256e465e&app_key=9d6e13e21b7e4031bdd230f16cc3bf6b


Gets air quality data feed

GET	/AirQuality
Test this endpoint

In [4]:
# URL
domain= "https://api.tfl.gov.uk"
directory ="/AirQuality"
url = domain+directory
print(url)

https://api.tfl.gov.uk/AirQuality


In [5]:
# We send the request to the API
# NOTE: if you don't have your APP_KEY, run this without the url_append
res = re.get((url)+url_append)

In [6]:
# We can check if the request was successful
res.status_code

200

In [8]:
# We can export the information that was returned using method .json()
res.json()

{'$id': '1',
 '$type': 'Tfl.Api.Presentation.Entities.LondonAirForecast, Tfl.Api.Presentation.Entities',
 'updatePeriod': 'hourly',
 'updateFrequency': '1',
 'forecastURL': 'http://londonair.org.uk/forecast',
 'disclaimerText': 'This forecast is intended to provide information on expected pollution levels in areas of significant public exposure. It may not apply in very specific locations close to unusually strong or short-lived local sources of pollution.',
 'currentForecast': [{'$id': '2',
   '$type': 'Tfl.Api.Presentation.Entities.CurrentForecast, Tfl.Api.Presentation.Entities',
   'forecastType': 'Current',
   'forecastID': '33405',
   'forecastBand': 'Low',
   'forecastSummary': 'Low air pollution forecast valid from Friday 17 September to end of Friday 17 September GMT',
   'nO2Band': 'Low',
   'o3Band': 'Low',
   'pM10Band': 'Low',
   'pM25Band': 'Low',
   'sO2Band': 'Low',
   'forecastText': 'A breezier day for Friday to give good dispersion of local emissions. For most of the 

In [9]:
thepayload=res.json()

In [10]:

#infoabout(thepayload)



NameError: name 'infoabout' is not defined

In [8]:


"Play with the data load to understand it"
def infoabout(thepayload):
    print("====================================")
    print("What's the object type")
    print(type(thepayload))
    print("====================================")
    print("Number of elements inside")
    print(len(thepayload))
    print("====================================")
    print("What are those elements")
    
    for count, value in enumerate(thepayload):
        test = value
        print(type(test),count, value)
        print("-----") 
    # print(count, value)

    #for x in thepayload:
    #    print(x)
    #    print("-----")
    









In [103]:
infoabout(thepayload)

What's the object type
<class 'list'>
Number of elements inside
786
What are those elements
<class 'dict'> 0 {'$type': 'Tfl.Api.Presentation.Entities.Place, Tfl.Api.Presentation.Entities', 'id': 'BikePoints_207', 'url': '/Place/BikePoints_207', 'commonName': 'Grosvenor Crescent, Belgravia', 'placeType': 'BikePoint', 'additionalProperties': [{'$type': 'Tfl.Api.Presentation.Entities.AdditionalProperties, Tfl.Api.Presentation.Entities', 'category': 'Description', 'key': 'TerminalName', 'sourceSystemKey': 'BikePoints', 'value': '001169', 'modified': '2021-09-16T23:11:56.97Z'}, {'$type': 'Tfl.Api.Presentation.Entities.AdditionalProperties, Tfl.Api.Presentation.Entities', 'category': 'Description', 'key': 'Installed', 'sourceSystemKey': 'BikePoints', 'value': 'true', 'modified': '2021-09-16T23:11:56.97Z'}, {'$type': 'Tfl.Api.Presentation.Entities.AdditionalProperties, Tfl.Api.Presentation.Entities', 'category': 'Description', 'key': 'Locked', 'sourceSystemKey': 'BikePoints', 'value': 'false'

In [15]:
key = input("PLEASE TYPE IN THE KEY LISTED ABOVE THAT YOUD LIKE TO EXPLORE : ")
print("====================================")

print("Print content of the key you just entered: -->" + key +"<--")
print("")
zoomofkey=thepayload[key]
infoabout(zoomofkey)
#print("It is a :" + str(type(zoomofkey)))
#print("THE TYPE of the object being returned from the value of key")

#print("Len of : "+str(len(zoomofkey)))

#print("content of ['" + key+"]' is ")
#print("====================================")

PLEASE TYPE IN THE KEY LISTED ABOVE THAT YOUD LIKE TO EXPLORE :  currentForecast


Print content of the key you just entered: -->currentForecast<--

What's the object type
<class 'list'>
Number of elements inside
2
What are those elements
<class 'dict'> 0 {'$id': '2', '$type': 'Tfl.Api.Presentation.Entities.CurrentForecast, Tfl.Api.Presentation.Entities', 'forecastType': 'Current', 'forecastID': '33405', 'forecastBand': 'Low', 'forecastSummary': 'Low air pollution forecast valid from Friday 17 September to end of Friday 17 September GMT', 'nO2Band': 'Low', 'o3Band': 'Low', 'pM10Band': 'Low', 'pM25Band': 'Low', 'sO2Band': 'Low', 'forecastText': 'A breezier day for Friday to give good dispersion of local emissions. For most of the day, the wind will be from the south-west, thereby bringing a &#39;clean&#39; feed of maritime air. A change in wind direction to south-easterly during the evening may import some particulates from the Paris area but this is not expected to be marked.&lt;br/&gt;&lt;br/&gt;Air pollution is expected to remain &#39;Low&#39; during the forecast p

In [54]:

print("we want to know the forecastType")
future = input("Value - Forecast for Current or Future ?")
#If I had more time, I would do  some some error handling .

#if I had more time I'd put a message that I didnt find the thing
def zoomin(this,value):
    for element in this:
        if element.get(value) == "Future":
            print("Okay, zooming into : " + (element.get(value)))
            return element
            
            


we want to know the forecastType


Value - Forecast for Current or Future ? 0


In [ ]:
theone=zoomin(zoomofkey,"forecastType")

In [31]:
infoabout(theone)


What's the object type
<class 'dict'>
Number of elements inside
12
What are those elements
<class 'str'> 0 $id
-----
<class 'str'> 1 $type
-----
<class 'str'> 2 forecastType
-----
<class 'str'> 3 forecastID
-----
<class 'str'> 4 forecastBand
-----
<class 'str'> 5 forecastSummary
-----
<class 'str'> 6 nO2Band
-----
<class 'str'> 7 o3Band
-----
<class 'str'> 8 pM10Band
-----
<class 'str'> 9 pM25Band
-----
<class 'str'> 10 sO2Band
-----
<class 'str'> 11 forecastText
-----


## Task
Parse the dictionary and print the AirQuality predictions for tomorrow

In [28]:
output = input("forecastSummary or forecastText")
print(theone[output])


forecastSummary or forecastText forecastText


Partly sunny with showers on Saturday.&lt;br/&gt;&lt;br/&gt;A light south or south easterly breeze is expected. This airflow will have travelled over the near continent and at times will have passed over the Paris region. It&#39;s likely to contain some particulates and as a result both PM10 &amp; PM2.5 particulates may increase during the day. It&#39;s not expected currently that particulate concentrations will reach the &#39;moderate&#39; pollution threshold.&lt;br/&gt;&lt;br/&gt;Air pollution is expected to remain &#39;Low&#39; during the forecast period for the following pollutants.&lt;br/&gt;&lt;br/&gt;Nitrogen Dioxide&lt;br/&gt;Sulphur Dioxide&lt;br/&gt;Ozone&lt;br/&gt;PM10 Particulates&lt;br/&gt;PM2.5 Particulates&lt;br/&gt;


make functions for unpacking nested lists and dictionaries
if I had more time, I would prolerly clean the URL 
try to make the code more pythonic... 
use collections or named tuples ?  or something more fancy.


## Task
What are the different modes of transport which are operated by Transfer for London? How many of modes do they have?

Print the list with different modes of transport, plus their count. Example output:
```
[bus, cable-car,.....]
Number of different modes of transport is: xyz
```

We need to search the documentation for correct request.

In [36]:
#counter ? 
# URL

directory ="/Mode/ActiveServiceTypes"
url = domain+directory
print(url+url_append)
# We send the request to the API
# NOTE: if you don't have your APP_KEY, run this without the url_append
res = re.get((url)+url_append)
res.status_code # would need to manage error handling here. 


https://api.tfl.gov.uk/Mode/ActiveServiceTypes?app_id=b09abd2194b44c4a8f80e810256e465e&app_key=9d6e13e21b7e4031bdd230f16cc3bf6b


200

In [43]:

# We can export the information that was returned using method .json()
layer1=res.json()
infoabout(layer1)



What's the object type
<class 'list'>
Number of elements inside
1
What are those elements
<class 'dict'> 0 {'$type': 'Tfl.Api.Presentation.Entities.ActiveServiceType, Tfl.Api.Presentation.Entities', 'mode': 'tube', 'serviceType': 'Regular'}
-----


## Task
How many BikePoints in London are operated by Transfor for London? How many docks are in **all** BikePoints? There is the information for empty and full docks for each BikePoint.

Answer : 
BikePoints: 786
Docks :20860

API USED 
https://api-portal.tfl.gov.uk/api-details#api=BikePoint&operation=BikePoint_GetAll&definition=Tfl.Api.Presentation.Entities.AdditionalProperties

In [134]:
# URL
domain= "https://api.tfl.gov.uk"
directory ="/BikePoint"
url = domain+directory
# We send the request to the API
# NOTE: if you don't have your APP_KEY, run this without the url_append
res = re.get((url)+url_append)

# We can check if the request was successful
res.status_code

# We can export the information that was returned using method .json()
#res.json()
thepayload=res.json()
#print(type(thepayload))
#infoabout(thepayload)

# value looks like something to exploit
BikePoints = set([])
for each in range(len(thepayload)):
    #print(thepayload[each])
    #print("--- \n")
    BikePoints.add(thepayload[each]["id"])
    
print("How many docks are in **all** BikePoints "+str(len(BikePoints)))
#print(BikePoints)
print(thepayload[0])

How many docks are in **all** BikePoints 786
{'$type': 'Tfl.Api.Presentation.Entities.Place, Tfl.Api.Presentation.Entities', 'id': 'BikePoints_207', 'url': '/Place/BikePoints_207', 'commonName': 'Grosvenor Crescent, Belgravia', 'placeType': 'BikePoint', 'additionalProperties': [{'$type': 'Tfl.Api.Presentation.Entities.AdditionalProperties, Tfl.Api.Presentation.Entities', 'category': 'Description', 'key': 'TerminalName', 'sourceSystemKey': 'BikePoints', 'value': '001169', 'modified': '2021-09-17T15:01:50.95Z'}, {'$type': 'Tfl.Api.Presentation.Entities.AdditionalProperties, Tfl.Api.Presentation.Entities', 'category': 'Description', 'key': 'Installed', 'sourceSystemKey': 'BikePoints', 'value': 'true', 'modified': '2021-09-17T15:01:50.95Z'}, {'$type': 'Tfl.Api.Presentation.Entities.AdditionalProperties, Tfl.Api.Presentation.Entities', 'category': 'Description', 'key': 'Locked', 'sourceSystemKey': 'BikePoints', 'value': 'false', 'modified': '2021-09-17T15:01:50.95Z'}, {'$type': 'Tfl.Api.Pre

In [132]:
layer1=thepayload[0]
infoabout(layer1["additionalProperties"])



What's the object type
<class 'list'>
Number of elements inside
9
What are those elements
<class 'dict'> 0 {'$type': 'Tfl.Api.Presentation.Entities.AdditionalProperties, Tfl.Api.Presentation.Entities', 'category': 'Description', 'key': 'TerminalName', 'sourceSystemKey': 'BikePoints', 'value': '001169', 'modified': '2021-09-17T15:01:50.95Z'}
-----
<class 'dict'> 1 {'$type': 'Tfl.Api.Presentation.Entities.AdditionalProperties, Tfl.Api.Presentation.Entities', 'category': 'Description', 'key': 'Installed', 'sourceSystemKey': 'BikePoints', 'value': 'true', 'modified': '2021-09-17T15:01:50.95Z'}
-----
<class 'dict'> 2 {'$type': 'Tfl.Api.Presentation.Entities.AdditionalProperties, Tfl.Api.Presentation.Entities', 'category': 'Description', 'key': 'Locked', 'sourceSystemKey': 'BikePoints', 'value': 'false', 'modified': '2021-09-17T15:01:50.95Z'}
-----
<class 'dict'> 3 {'$type': 'Tfl.Api.Presentation.Entities.AdditionalProperties, Tfl.Api.Presentation.Entities', 'category': 'Description', 'key':

In [69]:
infoabout(layer1["additionalProperties"][0])
layer2=layer1["additionalProperties"][0]

# why this doesnt work
#infoabout(layer1["additionalProperties"][0]["key"])
#but this one works 
#layer2=layer1["additionalProperties"][0]
print(layer2["key"])

#summary 


What's the object type
<class 'dict'>
Number of elements inside
6
What are those elements
<class 'str'> 0 $type
-----
<class 'str'> 1 category
-----
<class 'str'> 2 key
-----
<class 'str'> 3 sourceSystemKey
-----
<class 'str'> 4 value
-----
<class 'str'> 5 modified
-----
TerminalName


In [55]:
#GALOREEEEE 
for each in thepayload[0]:
    print(each)
    print(type(each))
   
        for properties in each:
        print(each("additionalProperties"))
        
#json to thepayload as a list  thepayload[0]
# thepayload[0] as a list that becomes layer1
#layer1["additionalProperties"] shows a list



$type
<class 'str'>
<class 'str'>


TypeError: 'str' object is not callable

https://api-portal.tfl.gov.uk/api-details#api=BikePoint&operation=BikePoint_GetAll&definition=Tfl.Api.Presentation.Entities.AdditionalProperties


In [102]:
NbDocks=0
#chainmap ? ?

for x in range(len(thepayload)):
    #print(thepayload[x])
#     print(type(thepayload[x]))
#     a=thepayload[x]
#     print(a)
#     print(type(a)) 
    #print(thepayload[x]["additionalProperties"][0]["key"])
        for properties in thepayload[x]["additionalProperties"]:
            key,value=properties["key"],properties["value"]
            #print(key + " " + value)
            if properties["key"]== "NbDocks":
                NbDock = int(properties["value"])
               # print(NbDock)
                NbDocks=NbDocks+NbDock
                
            

        
print("How many docks are in all of london? :" + str(NbDocks))

How many docks are in all BikePoints? :20860


In [165]:
#zzzz
#earlier versions 
def newzoominthisdic(layer,key,value):
    numdock=0   
    for i in layer:
        if i.get(key)== value:
            print(i.get("value"))
            numdock+=int(i.get("value"))
            
newzoominthisdic(layer2,"key","NbDocks")


18


## Task
How many tube and bus lines are in London? Print names of all tube lines.

In [ ]:
#Cannot find the lists of bus tubes lines all whole of london


## Task
How many station has `victoria` line? 
Answer : 16

In [129]:
# URL https://api.tfl.gov.uk/Line/victoria/Arrivals
domain= "https://api.tfl.gov.uk"
directory ="/Line/victoria/Arrivals"
url = domain+directory
# We send the request to the API
# NOTE: if you don't have your APP_KEY, run this without the url_append
res = re.get((url)+url_append)

# We can check if the request was successful
res.status_code

# We can export the information that was returned using method .json()
#res.json()
thepayload=res.json()
print(type(thepayload))
#infoabout(thepayload)
  
#print(BikePoints)

#print(thepayload[0])

<class 'list'>


In [128]:
surfaceelement=thepayload[0]
print(type(surfaceelement))

Stations = set([])
for station in range(len(thepayload)):
    #print(thepayload[each])
    #print("--- \n")
    Stations.add(thepayload[station]["stationName"])
    
print("How many station has victoria line? "+str(len(Stations)))


<class 'dict'>
How many station has victoria line? 16


## Task
Plan the journey from Heathrow Airport to Tower Bridge using Bus and Tube? Which way is faster? Example output:
```
Planned duration:
Bus: x minutes
Tube: y minutes
```

We need to search the documentation for correct requests and parameters we need.